### ISSUE TITLE
Optional properties on `.with(...)` narrowed type missing

### ISSUE THREAD

**Issue by julienavert** [gvergnaud/ts-pattern#142]
Hey, thanks for the amazing library!

I think this issue is related to #20, but the initial setup is slightly different. I commented on #20, but since it is closed, I am afraid it won't notify anyone.

**Describe the bug**

When the generic type used in `match()` contains optional properties, if they are not part of the matching pattern, they get dropped from the narrowed down type available to the handler in `.with(...)`.

**Code Sandbox with a minimal reproduction case**

Sorry I used typescriptlang playground instead (probably OK since it is only a typings issue).

[Here is a code example](https://www.typescriptlang.org/play?target=99&module=1&pretty=true&allowSyntheticDefaultImports=true&ts=4.7.4#code/JYWwDg9gTgLgBAbziAhjAxgCzgXzgMyghDgHIYBnAWjDRgFMoA7UgKFfqYFcSBlY+gFFuJBKzgS4AMQgQ4AXjIyIpADTjJAIRRQFZbVDY52MAJ5h6cfiHoB5AEYAreuniKxkuCgD8ALisCwjwA3BoS9n5wFDBQwEwA5qGe6JH2sgA29ChMocas6BBM0XBxYFww-tZ2Ti5uiGFelYEiAHTK6p72-qRMEPCgkLDZMEbsBUXwUPQUXOl1qBiYABSl5QCUDS0A7sAwy0goTTZBIG2yuKpwAG4o6VyW8gB8cAAGyv4AJAg3d-Qt9jgXhtPNtdvtGgFjq0DBdrrd7gpni8DJ9vvC-gCgZsdnslgd-FwmAATej4OL0Imwn4Ip6vAA8hJJZKYFMeqOpGMBwMkLXoAA9MCguNFgFd6Es1qF8oUKBBMi10hB4kspjM5pLWEA)

inlined:

```ts
enum SomeEnum {
    Foo = 'Foo',
    Bar = 'Bar'
}

type SomeObject = {
    a?: SomeEnum;
    b?: string;
    c?: boolean;
}

const input: SomeObject = {
    a: SomeEnum.Foo,
    b: 'not important'
}

const result = match(input)
    .with({ a: SomeEnum.Foo }, value => `Foo: ${value.b}`)
    .with({ a: SomeEnum.Bar }, value => `Bar: ${value.b}`)
    .with({ a: undefined }, value => `<undefined>: ${value.b}`)
    .exhaustive();
```

In the playground you can see that the `with` narrowing is removing all optional properties from the input type

![image](https://user-images.githubusercontent.com/87069068/216660721-b7442916-6748-4f25-be3c-f8c7bdb8d35c.png)

What I would have expected:

```ts
{
    a: SomeEnum.Foo;
    b?: string;
    c?: boolean;
}
```

or maybe even

```ts
{
    a: SomeEnum.Foo;
    b: string | undefined;
    c: boolean | undefined;
}
```

It is interesting to note that if the initial type for the property `b` is `string | undefined` instead of the property being optional, it works as expected (it even fixes the issue for **other** properties!) [typescriptlang playground](https://www.typescriptlang.org/play?target=99&module=1&pretty=true&allowSyntheticDefaultImports=true&ts=4.7.4#code/JYWwDg9gTgLgBAbziAhjAxgCzgXzgMyghDgHIYBnAWjDRgFMoA7UgKFfqYFcSBlY+gFFuJBKzgS4AMQgQ4AXjIyIpADTjJAIRRQFZbVDY52MAJ5h6cfiHoB5AEYAreuniKxkuCgD8ALisCwjwA3BoS9v4UMFDATADmcAA+cFxMACb0+LH0aaGe6H5w9rIANvQoTKHGrOgQTFFwsWBcMP7Wdk4ubohhXm2BIgB0yuqeEWRMEPCgkLAVMEbstfXwUPQUXCXdqBiYABRNLQCUvYMA7sAw+0go-TZBIMOyuKpwAG4oJVyW8gB8cAADZT+AAkCA+X3og3sOABJ0850u1z6AXuQwML3en2+Cn+AIMoPB2KhMLhpwuVz2N38qQyWSYOUxEJxf0BAB5aZlsmlfoTmSTYfDJIN6AAPTAoLhRYBveh7I6hGp1CgQMqDEoQOJ7NYbLYK1hAA)

inlined:

```ts
enum SomeEnum {
    Foo = 'Foo',
    Bar = 'Bar'
}

type SomeObject = {
    a?: SomeEnum;
    b: string | undefined;
    c?: boolean;
}

const input: SomeObject = {
    a: SomeEnum.Foo,
    b: 'not important'
}

const result = match(input)
    .with({ a: SomeEnum.Foo }, value => `Foo: ${value.b}`)
    .with({ a: SomeEnum.Bar }, value => `Bar: ${value.b}`)
    .with({ a: undefined }, value => `<undefined>: ${value.b}`)
    .exhaustive();
```

![image](https://user-images.githubusercontent.com/87069068/216661819-4a9292d4-6145-4e7a-802f-7a6b3bc9b4fa.png)

**Versions**
- TypeScript version: 4.7.4
- ts-pattern version: 4.1.3

**Comment by gvergnaud** [comment]
Thanks for the report, I'll take a look

**Comment by bonqus** [comment]
I experience the same problem, `when` gives the correct type.

**Comment by gvergnaud** [comment]
Thanks, fixed it in v 4.2.2 

[Playground](https://www.typescriptlang.org/play?target=99&module=1&pretty=true&allowSyntheticDefaultImports=true&ts=4.7.4&ssl=3&ssc=1&pln=23&pc=19#code/JYWwDg9gTgLgBAbziAhjAxgCzgXzgMyghDgHIYBnAWjDRgFMoA7UgKFfqYFcSBlY+gFFuJBKzgS4AMQgQ4AXjIyIpADTjJAIRRQFZbVDY52MAJ5h6cfiHoB5AEYAreuniKxkuCgD8ALisCwjwA3BoS9n5wFDBQwEwA5qGe6JH2sgA29ChMocas6BBM0XBxYFww-tZ2Ti5uiGFelYEiAHTK6p72-qRMEPCgkLDZMEbsBUXwUPQUXOl1qBiYABSl5QCUDS0A7sAwy0goTTZBIG2yuKpwAG4o6VyW8gB8cAAGyv4AJAg3d-Qt9jgXhtPNtdvtGgFjq0DBdrrd7gpni8DJ9vvC-gCgZsdnslgd-FwmAATej4OL0Imwn4Ip6vAA8hJJZKYFMeqOpGMBwMkLXoAA9MCguNFgFd6Es1qF8oUKBBMi10hB4kspjM5pLWEA)

Release note: https://github.com/gvergnaud/ts-pattern/releases/tag/v4.2.2
